docs/designs/sqlflow_integration_design.md

```diff
@@ -0,0 +1,237 @@
+            features = tf.concat(features, -1)
+            return tf.reshape(features, features_shape)
+
+        record = tf.strings.to_number(record, tf.float32)
+        features_shape = ({{.InputShape}}, 1)
+        labels_shape = (1,)
+        {{if .IsTraining}}
+        label_col_name = "{{.LabelColName}}"
+        if mode != Mode.PREDICTION:
+            if label_col_name not in metadata.column_names:
+                raise ValueError(
+                    "Missing the label column '%s' in the retrieved "
+                    "table." % label_col_name
+                )
+            label_col_ind = metadata.column_names.index(label_col_name)
+            labels = tf.reshape(record[label_col_ind], labels_shape)
+            return (
+                _get_features_without_labels(
+                    record, label_col_ind, features_shape
+                ),
+                labels,
+            )
+        {{end}}
+        return tf.reshape(record, features_shape)
+
+    dataset = dataset.map(_parse_data)
+
+    {{if .IsTraining}}
+    if mode != Mode.PREDICTION and "{{.TrainClause.EnableShuffle}}" == "true":
+        dataset = dataset.shuffle(buffer_size={{.TrainClause.ShuffleBufferSize}})
+    {{end}}
+
+    return dataset
+```
+
+Some fields used to generate the above `dataset_fn` are obtained directly from the extended SQL statement. For example, ``FeaturesList`` is obtained
+from `SELECT FROM` clause. `LabelColName` is obtained from `LABEL` clause. `TrainClause.ShuffleBufferSize` is obtained from
+`train.shuffle` in the `WITH` clause. There are also fields that are obtained indirectly. For example, `InputShape` is inferred from `FeaturesList`.
+
+Note that in the template we currently we hard-coded the types for each column to be ``tf.float32`` in the generated `dataset_fn`. We should infer
+this information from the database instead. We also hard-coded other components in the model definition such as ``loss`` and ``optimizer``, these
+components should be derived from the model zoo instead.
+
+#### Generate ElasticDL Command
+
+Once we generated the components for the model definition, we can then generate the ElasticDL command to submit the job. 
+Below is an example:
+
+```sh
+elasticdl train \
+--image_base=elasticdl:ci \
+--model_zoo=<model-zoo> \
+--model_def=<path-to-generated-model-def> \
+--loss=<loss-function-name> \
+--eval_metrics_fn=<eval-metrics-function-name> \
+--training_data=<training-input-table> \
+--evaluation_data=<evaluation-input-table> \
+--num_epochs=2 \
+--master_resource_request="cpu=400m,memory=1024Mi" \
+--master_resource_limit="cpu=1,memory=2048Mi" \
+--worker_resource_request="cpu=400m,memory=2048Mi" \
+--worker_resource_limit="cpu=1,memory=3072Mi" \
+--minibatch_size=64 \
+--num_minibatches_per_task=10 \
+--num_workers=2 \
+--checkpoint_steps=10 \
+--evaluation_steps=15 \
+--grads_to_wait=2 \
+--job_name=test-iris \
+--log_level=INFO \
+--image_pull_policy=Never \
+--output=<model-output> \
```

Comment:
Looks like we save the trained model at `--output`. Is `--output` determined by the `INTO` clause in the SQL? If so, please be aware that `TrainIR` doesn't contain the `INTO` clause due to [this reason](https://github.com/sql-machine-learning/sqlflow/blob/1fc8e5d31261c16d74416922dac5689c42f83f9f/pkg/sql/codegen/intermediate_representation.go#L64-L66).

Reply:
"The sql package will save the output files of a generated Python program." -> how are these files saved? So where else would I obtain this `INTO` clause if it's not in `TrainIR`?

Reply:
> how are these files saved?

In a training job, SQLFlow runs the EDL submitter program in a temporary directory. When training is finished, SQLFlow serialized the directory via zip, and dump the zip file to an ODPS table specified by the `INTO` clause.

> So where else would I obtain this `INTO` clause if it't not in `TrainIR`?

So, I believe EDL shouldn't know `INTO` clause.